app/models/sorn_xml_parser.rb

```diff
@@ -1,95 +1,127 @@
+      contesting: find_section('CONTESTING'),
+      notification: find_section('NOTIFICATION'),
+      exemptions: find_section('EXEMPTIONS'),
+      history: find_section('HISTORY'),
+      headers: @sections.keys }
   end
 
-  def get_agency
-    find_tag('AGENCY')
+  def get_supplementary_information
+    content = @parser.within('SUPLINF').map{ |node| node unless unwanted_parts_of_suplinf(node) }
+    cleanup_xml_element_to_string(content)
   end
 
-  def get_summary
-    find_tag('SUM')
+  def get_system_name
+    @system_name = strip_tags(find_section('SYSTEM NAME'))
   end
 
-  def get_addresses
-    find_tag('ADD')
+  def get_system_number
+    number = strip_tags(find_section('NUMBER'))
+    # puts number
+    if number and @system_name
+      parse_system_name_from_number
+    end
   end
 
-  def get_further_information
-    find_tag('FURINF')
+  def find_tag(tag)
+    element = @parser.for_tag(tag).first
+    cleanup_xml_element_to_string(element)
   end
 
-  def get_supplementary_information
-    # custom to get everything but priact and sig
-    # an example of why paragraphs only won't work
-    # https://www.federalregister.gov/documents/full_text/xml/2020/10/13/2020-22534.xml
-    content = []
-    @parser.within('SUPLINF').each do |node|
-      if node.name != 'PRIACT' && node.name != 'SIG'
-        # skip section title
-        if node.name == 'HD' && node.include?('SUPPLEMENTARY')
-          next
-        else
-          content << node
-        end
+  def find_section(header)
+    # Get a named section of the PRIACT tag
+    # header of 'NUMBER' will match the section with key 'System Name and Number'
+    matched_header = @sections.keys.find{ |key| key.upcase.include? header }
+    @sections[matched_header]
+  end
+
+  private
+
+  def get_sections
+    # Gather the named sections of the PRIACT tag
+    sections = {}
+    current_header = nil
+    @parser.within('PRIACT').each do |node|
+      if node.name == 'HD'
+        current_header = cleanup_xml_element_to_string(node)
+        sections[current_header] = []
+      elsif current_header.nil?
+        next
+      elsif node.name == 'P'
+        # append cleaned strings
+        sections[current_header] << cleanup_xml_element_to_string(node)
       end
     end
-    content
+
+    # discard the rare nil keys
+    sections.except!(nil)
+    # Join array into a large string of paragraphs.
+    sections.transform_values!{|values| values.map{|p| "<p>#{p}</p>" }.join(" ") }
   end
 
-  def get_system_name
-    bracketed_name = find_section('SYSTEM NAME')
-    if bracketed_name
-      @system_name = bracketed_name.join(', ')
-      @system_name = @system_name.sub ', {}', ''
-      @system_name = @system_name.sub '"]', ''
-      @system_name = @system_name.sub '["', ''
-      @system_name = @system_name.sub '"', ''
-      return @system_name
-    end
+  def cleanup_xml_element_to_string(element)
+    # recursively convert hashes and array down to a string
```

Comment:
Yep. This is the main part I want to pair with you on. When our XML gets parsed it can have elements that are hashes, arrays, or strings. They can contain nested hashes and arrays to any depth. This method recursively pulls all the strings out of anything it finds.

Reply:
I tried a case statement here instead too. I think it is harder to read.
```
    case element.class.to_s
    when 'Hash', 'Saxerator::Builder::HashElement'
      cleanup_xml_element_to_string(xml_hash_to_string(element))
    when 'Array', 'Saxerator::Builder::ArrayElement'
      cleanup_xml_element_to_string(xml_array_to_string(element))
    when 'String', 'Saxerator::Builder::StringElement'
      element.strip
    end
```